File: C++/FourSum.cpp
#include<iostream>
#include<stdlib.h>
#include<vector>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;
class Solution {
public:
	vector<vector<int> > fourSum(vector<int> &num, int target) {
		// Start typing your C/C++ solution below
		// DO NOT write int main() function
		vector<vector<int>> ret;
		if(num.size() < 4)
			return ret;
		sort(num.begin(), num.end());
		for(int i = 0; i < num.size()-3; i++)
		{
			if(i != 0 && num[i] == num[i-1])
				continue;
			for(int j = i + 1; j < num.size() - 2; j++)
			{
				if(j != (i + 1) && num[j] == num[j-1])
					continue;
				int k = j + 1;
				int m = num.size()-1;
				while(k < m)
				{
					if(k != (j+1) && num[k] == num[k-1])
					{
						k++;
						continue;
					}
					if(m != (num.size()-1) && num[m] == num[m+1])
					{
						m--;
						continue;
					}
					int sum = num[i] + num[j] + num[k] + num[m];
					if(sum == target)
					{
						vector<int> list;
						list.push_back(num[i]);
						list.push_back(num[j]);
						list.push_back(num[k]);
						list.push_back(num[m]);
						ret.push_back(list);
						k++;
					}
					else if(sum < target)
						k++;
					else if(sum > target)
						m--;
				}
			}
		}
		return ret;
	}
};

File: JAVA/FourSum.java
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
	ArrayList<ArrayList<Integer>> ret = new ArrayList<ArrayList<Integer>>();
    public ArrayList<ArrayList<Integer>> fourSum(int[] num, int target) {
        // Start typing your Java solution below
        // DO NOT write main() function
    	ret.clear();
    	if(num == null || num.length < 4)
    		return ret;
    	Arrays.sort(num);
        int len = num.length;
        for(int i = 0; i < len ;i++)
        {
        	if(i > 0 && num[i] == num[i-1])
        		continue;
        	if(num[i] > 0 &&  num[i] > target)
        		break;
        	if(4*num[i] > target)
        		break;
        	for(int j = i+1; j < len; j++)
        	{
        		if(j > i+1 && num[j] == num[j-1])
        			continue;
        		if(3*num[j] + num[i] > target)
        			break;
        		int k = j+1;
        		int m = len-1;
        		while(k < m)
        		{
        			if(k > j+1 && num[k] == num[k-1])
        			{
        				k++;
        				continue;
        			}
        			if(m < len-1 && num[m] == num[m+1])
        			{
        				m--;
        				continue;
        			}
        			int sum = num[i] + num[j] + num[k] + num[m];
        			if(sum == target)
        			{
        				ArrayList<Integer> cur = new ArrayList<Integer>();
        				cur.add(num[i]);
        				cur.add(num[j]);
        				cur.add(num[k]);
        				cur.add(num[m]);
        				ret.add(cur);
        				k++;
        			}
        			else if(sum < target)
        				k++;
        			else
        				m--;
        		}
        	}
        }
        return ret;
    }
}
public class FourSum {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		//[-1,2,2,-5,0,-1,4], 3
		//[-1,-5,-5,-3,2,5,0,4], -7
		//[1,-2,-5,-4,-3,3,3,5], -11
		//[-7,0,-5,0,-8,-3,-5,1,4,-2,-6,10], -12  [-8,-6,-2,4]
		int[] num = {-7,0,-5,0,-8,-3,-5,1,4,-2,-6,10};
		int target = -12;
		Solution sol = new Solution();
		System.out.println(sol.fourSum(num, target));
	}

}
